File: languages/go/entity/record/record_details.go
package record

import (
	"github.com/bloock/bloock-sdk-go/v2/entity/authenticity"
	"github.com/bloock/bloock-sdk-go/v2/entity/integrity"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
)

// IntegrityDetails represents details related to the integrity of a record, including hash and proof.
type IntegrityDetails struct {
	Hash  string
	Proof *integrity.Proof
}

func NewIntegrityDetailsFromProto(r *proto.IntegrityDetails) *IntegrityDetails {
	if r == nil {
		return nil
	}

	integrityDetails := IntegrityDetails{
		Hash:  r.Hash,
		Proof: nil,
	}

	if r.Proof != nil {
		proof := integrity.NewProofFromProto(r.Proof)
		integrityDetails.Proof = &proof
	}

	return &integrityDetails
}

func (r *IntegrityDetails) ToProto() *proto.IntegrityDetails {
	return &proto.IntegrityDetails{
		Hash:  r.Hash,
		Proof: r.Proof.ToProto(),
	}
}

// AuthenticityDetails represents details related to the authenticity of a record, including signatures.
type AuthenticityDetails struct {
	Signatures []authenticity.Signature
}

func NewAuthenticityDetailsFromProto(r *proto.AuthenticityDetails) *AuthenticityDetails {
	if r == nil {
		return nil
	}

	signatures := make([]authenticity.Signature, len(r.Signatures))
	for i, signature := range r.Signatures {
		signatures[i] = authenticity.NewSignatureFromProto(signature)
	}

	return &AuthenticityDetails{
		Signatures: signatures,
	}
}

func (r *AuthenticityDetails) ToProto() *proto.AuthenticityDetails {
	signatures := make([]*proto.Signature, len(r.Signatures))
	for i, signature := range r.Signatures {
		signatures[i] = signature.ToProto()
	}

	return &proto.AuthenticityDetails{
		Signatures: signatures,
	}
}

// EncryptionDetails represents details related to the encryption of a record, including algorithm, key, and subject.
type EncryptionDetails struct {
	Alg     *string
	Key     *string
	Subject *string
}

func NewEncryptionDetailsFromProto(r *proto.EncryptionDetails) *EncryptionDetails {
	if r == nil {
		return nil
	}

	return &EncryptionDetails{
		Alg:     r.Alg,
		Key:     r.Key,
		Subject: r.Subject,
	}
}

func (r *EncryptionDetails) ToProto() *proto.EncryptionDetails {
	return &proto.EncryptionDetails{
		Alg:     r.Alg,
		Key:     r.Key,
		Subject: r.Subject,
	}
}

// AvailabilityDetails represents details related to the availability of a record, including content type and size.
type AvailabilityDetails struct {
	ContentType *string
	Size        int64
}

func NewAvailabilityDetailsFromProto(r *proto.AvailabilityDetails) *AvailabilityDetails {
	if r == nil {
		return nil
	}

	return &AvailabilityDetails{
		ContentType: r.Type,
		Size:        r.Size,
	}
}

func (r *AvailabilityDetails) ToProto() *proto.AvailabilityDetails {
	return &proto.AvailabilityDetails{
		Type: r.ContentType,
		Size: r.Size,
	}
}

// RecordDetails represents all details related to a record, including integrity, authenticity, encryption, and availability details.
type RecordDetails struct {
	IntegrityDetails    *IntegrityDetails
	AuthenticityDetails *AuthenticityDetails
	EncryptionDetails   *EncryptionDetails
	AvailabilityDetails *AvailabilityDetails
}

func NewRecordDetailsFromProto(r *proto.RecordDetails, configData *proto.ConfigData) RecordDetails {
	if r == nil {
		return RecordDetails{}
	}

	return RecordDetails{
		IntegrityDetails:    NewIntegrityDetailsFromProto(r.Integrity),
		AuthenticityDetails: NewAuthenticityDetailsFromProto(r.Authenticity),
		EncryptionDetails:   NewEncryptionDetailsFromProto(r.Encryption),
		AvailabilityDetails: NewAvailabilityDetailsFromProto(r.Availability),
	}
}

func (r *RecordDetails) ToProto() *proto.RecordDetails {
	return &proto.RecordDetails{
		Integrity:    r.IntegrityDetails.ToProto(),
		Authenticity: r.AuthenticityDetails.ToProto(),
		Encryption:   r.EncryptionDetails.ToProto(),
		Availability: r.AvailabilityDetails.ToProto(),
	}
}

File: languages/js/src/entity/record/record-details.ts
import * as proto from "../../bridge/proto/record_entities";
import { Proof } from "../integrity";
import { Signature } from "../authenticity";

/**
 * Represents details related to the integrity of a record, including hash and proof.
 */
export class IntegrityDetails {
  hash: string;
  proof?: Proof;

  /**
   * Constructs a IntegrityDetails object with the specified parameters.
   * @param hash 
   * @param proof 
   */
  constructor(hash: string, proof?: Proof) {
    this.hash = hash;
    this.proof = proof;
  }

  static fromProto(r: proto.IntegrityDetails) {
    return new IntegrityDetails(
      r.hash,
      r.proof ? Proof.fromProto(r.proof) : undefined
    );
  }

  toProto(): proto.IntegrityDetails {
    return proto.IntegrityDetails.fromPartial({
      hash: this.hash,
      proof: this.proof?.toProto()
    });
  }
}

/**
 * Represents details related to the authenticity of a record, including signatures.
 */
export class AuthenticityDetails {
  signatures: Signature[];

  /**
   * Constructs a AuthenticityDetails object with the specified parameters.
   * @param signatures 
   */
  constructor(signatures: Signature[]) {
    this.signatures = signatures;
  }

  static fromProto(r: proto.AuthenticityDetails) {
    return new AuthenticityDetails(
      r.signatures.map(s => Signature.fromProto(s))
    );
  }

  toProto(): proto.AuthenticityDetails {
    return proto.AuthenticityDetails.fromPartial({
      signatures: this.signatures.map(s => s.toProto())
    });
  }
}

/**
 * Represents details related to the encryption of a record, including algorithm, key, and subject.
 */
export class EncryptionDetails {
  alg?: string;
  key?: string;
  subject?: string;

  /**
   * Constructs a EncryptionDetails object with the specified parameters.
   * @param alg 
   * @param key 
   * @param subject 
   */
  constructor(alg?: string, key?: string, subject?: string) {
    this.alg = alg;
    this.key = key;
    this.subject = subject;
  }

  static fromProto(r: proto.EncryptionDetails) {
    return new EncryptionDetails(r.alg, r.key, r.subject);
  }

  toProto(): proto.EncryptionDetails {
    return proto.EncryptionDetails.fromPartial({
      alg: this.alg,
      key: this.key,
      subject: this.subject
    });
  }
}

/**
 * Represents details related to the availability of a record, including content type and size.
 */
export class AvailabilityDetails {
  contentType?: string;
  size: number;

  /**
   * Constructs a AvailabilityDetails object with the specified parameters.
   * @param size 
   * @param contentType 
   */
  constructor(size: number, contentType?: string) {
    this.size = size;
    this.contentType = contentType;
  }

  static fromProto(r: proto.AvailabilityDetails) {
    return new AvailabilityDetails(r.size, r.type);
  }

  toProto(): proto.AvailabilityDetails {
    return proto.AvailabilityDetails.fromPartial({
      type: this.contentType,
      size: this.size
    });
  }
}

/**
 * Represents all details related to a record, including integrity, authenticity, encryption, and availability details.
 */
export class RecordDetails {
  integrity?: IntegrityDetails;
  authenticity?: AuthenticityDetails;
  encryption?: EncryptionDetails;
  availability?: AvailabilityDetails;

  /**
   * Constructs a RecordDetails object with the specified parameters.
   * @param integrity 
   * @param authenticity 
   * @param encryption 
   * @param availability 
   */
  constructor(
    integrity?: IntegrityDetails,
    authenticity?: AuthenticityDetails,
    encryption?: EncryptionDetails,
    availability?: AvailabilityDetails
  ) {
    this.integrity = integrity;
    this.authenticity = authenticity;
    this.encryption = encryption;
    this.availability = availability;
  }

  static fromProto(r: proto.RecordDetails) {
    return new RecordDetails(
      r.integrity ? IntegrityDetails.fromProto(r.integrity) : undefined,
      r.authenticity
        ? AuthenticityDetails.fromProto(r.authenticity)
        : undefined,
      r.encryption ? EncryptionDetails.fromProto(r.encryption) : undefined,
      r.availability ? AvailabilityDetails.fromProto(r.availability) : undefined
    );
  }

  toProto(): proto.RecordDetails {
    return proto.RecordDetails.fromPartial({
      integrity: this.integrity?.toProto(),
      authenticity: this.authenticity?.toProto(),
      encryption: this.encryption?.toProto(),
      availability: this.availability?.toProto()
    });
  }
}
